### PR TITLE
Tweak "borrow closure argument" suggestion

### LINKED ISSUE #45727: customize error message in the case where user supplies erroneous type annotation on closure argument
As [discussed in https://github.com/rust-lang/rust/pull/45072](https://github.com/rust-lang/rust/pull/45072#issuecomment-341096796), we have a newer approach to handling type annotations on closure arguments that should allow us to give clearer errors in many cases. Currently, if the user annotations would yield a unification error, we wind up falling back to some more general code. So for example in the `ui/anonymous-higher-ranked-lifetime.rs` test, for code like:

```rust
fn g2<F>(_: F) where F: Fn(&(), fn(&())) {}

fn main() {
    g2(|_: (), _: ()| {});
}
```

we generate

```
[00:39:53] error[E0631]: type mismatch in closure arguments
[00:39:53]   --> $DIR/anonymous-higher-ranked-lifetime.rs:18:5
[00:39:53]    |
[00:39:53] 18 |     g2(|_: (), _: ()| {});
[00:39:53]    |     ^^ ----------------- found signature of `fn((), ()) -> _`
[00:39:53]    |     |
[00:39:53]    |     expected signature of `for<'r> fn(&'r (), for<'s> fn(&'s ())) -> _`
[00:39:53]    |
[00:39:53]    = note: required by `g2`
```

but we could now generate something specific to one of the parameter types, rather like:

```
[00:39:53] error[E0631]: type mismatch in closure argument
[00:39:53]   --> $DIR/anonymous-higher-ranked-lifetime.rs:18:5
[00:39:53]    |
[00:39:53] 18 |     g2(|_: (), _: ()| {});
[00:39:53]    |            ^^ argument needs type `&()`
[00:39:53]    |     
[00:39:53]    = note: expected type `&()`
[00:39:53]               found type `()`
```

To do this, we would want to search around for the `FIXME(#45727)` -- this identifies the newer code, which is currently running in a transaction and falling back to the older strategy in the event of error.

### PR DESCRIPTION
**Pull request by estebank**
Fix #45727.

### CODE REVIEW

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs [hunk 1]
```diff
@@ -3732,29 +3736,44 @@ fn hint_missing_borrow<'tcx>(
...
+                    ty = mut_ty.ty;
+                    left -= 1;
+                    info!(?span, ?left, "after");
+                }
+                info!(?span, ?left, "result");
+                if left == 0 {
+                    remove_borrow.push((span, String::new()));
+                } else {
+                    remove_borrow.push((arg.span, expected_arg.to_string()));
+                }
```
Comment (petrochenkov) [inline compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:3762]: Nit: the `remove_borrow.push(...)` part can be moved out of the condition.

File: tests/ui/closures/multiple-fn-bounds.stderr [hunk 2]
```diff
@@ -20,6 +18,10 @@ note: required by a bound in `foo`
    |
 LL | fn foo<F: Fn(&char) -> bool + Fn(char) -> bool>(f: F) {
    |                               ^^^^^^^^^^^^^^^^ required by this bound in `foo`
+help: do not borrow the argument
+   |
+LL |     foo(move |char| v);
+   |               ~~~~
```
Comment (estebank) [inline tests/ui/closures/multiple-fn-bounds.stderr:24]: I'm not *adding* this incorrect suggestion, just changing its presentation, but we need to fix this in a follow up.

**Comment by petrochenkov** [comment]
r=me after squashing commits.
@rustbot author

**Pull request merged**